contracts/Uniswap/UniswapAnchoredView.sol

```diff
@@ -0,0 +1,278 @@
+        }
+    }
+
+    /**
+     * @dev Fetches the current token/eth price accumulator from uniswap.
+     */
+    function currentCumulativePrice(TokenConfig memory config) internal view returns (uint) {
+        (uint cumulativePrice0, uint cumulativePrice1,) = UniswapV2OracleLibrary.currentCumulativePrices(config.uniswapMarket);
+        if (config.isUniswapReversed) {
+            return cumulativePrice1;
+        } else {
+            return cumulativePrice0;
+        }
+    }
+
+    /**
+     * @dev Fetches the current eth/usd price from unsiwap, with 6 decimals of precision.
+     *  Conversion factor is 1e18 for eth/usdc market, since we decode uniswap price statically with 18 decimals.
+     */
+    function fetchEthPrice() internal returns (uint) {
+        return fetchAnchorPrice(getTokenConfigBySymbolHash(ethHash), 1e18);
+    }
+
+    /**
+     * @dev Fetches the current token/usd price from uniswap, with 6 decimals of precision.
+     */
+    function fetchAnchorPrice(TokenConfig memory config, uint conversionFactor) internal virtual returns (uint) {
+        (uint nowCumulativePrice, uint oldCumulativePrice, uint oldTimestamp) = pokeWindowValues(config);
+        uint timeElapsed = block.timestamp - oldTimestamp;
+
+        // Calculate uniswap time-weighted average price
+        FixedPoint.uq112x112 memory priceAverage = FixedPoint.uq112x112(uint224((nowCumulativePrice - oldCumulativePrice) / timeElapsed));
+        uint anchorPriceUnscaled = mul(priceAverage.decode112with18(), conversionFactor);
+        uint anchorPrice;
+
+        // Adjust anchor price to val * 1e6 decimals format
+        if (config.isUniswapReversed) {
+            anchorPrice = anchorPriceUnscaled / config.baseUnit;
+        } else {
+            anchorPrice = mul(anchorPriceUnscaled, config.baseUnit) / 1e36;
+        }
+
+        emit AnchorPriceUpdate(config.uniswapMarket, anchorPrice, nowCumulativePrice, oldCumulativePrice, oldTimestamp);
+
+        return anchorPrice;
+    }
+
+    /**
+     * @dev Get time-weighted average prices for a token at the current timestamp.
+     *  Update new and old observations of lagging window if period elapsed.
+     */
+    function pokeWindowValues(TokenConfig memory config) internal returns (uint, uint, uint) {
+        address uniswapMarket = config.uniswapMarket;
+        uint cumulativePrice = currentCumulativePrice(config);
+
+        Observation storage newObservation = newObservations[uniswapMarket];
+        Observation storage oldObservation = oldObservations[uniswapMarket];
+
+        // Update new and old observations if elapsed time is bigger or equal to anchor period
+        uint timeElapsed = block.timestamp - newObservation.timestamp;
+        if (timeElapsed >= anchorPeriod) {
+            emit UniswapWindowUpdate(uniswapMarket, oldObservation.timestamp, newObservation.timestamp, oldObservation.acc, newObservation.acc);
+            oldObservation.timestamp = newObservation.timestamp;
```

Comment:
actually, since Observation is storage and not memory, are we incurring SLOAD every time we reference property here?

Reply:
fixed in #101